### PR TITLE
Multi edge support for igraph_bipartite_game_gnm

### LINKED ISSUE #2374: igraph_bipartite_game_gnm() should support multi-edges
**What is the feature or improvement you would like to see?**

`igraph_bipartite_game_gnm()` should be capable of generating multi-edges. This is analogous to #2161.

**Use cases for the feature**

 - Analogy with `igraph_erdos_renyi_game_gnm()`
 - Ability to generate random bipartite multigraphs

**References**

 - #2161

### PR DESCRIPTION
**Pull request by RohithS98**
New feature: Multi edge support for `igraph_bipartite_game_gnm` .

Resolves #2374 

I am working on adding test cases.

### CODE REVIEW

**Comment by szhorvat** [comment]
I'm sorry, I should have mention that this change must be based on the `develop` branch, as it is breaking. It will be included in version 0.11, not the next 0.10.x version.

Can you rebase it?

**Comment by szhorvat** [comment]
Great work so far! Please don't put effort into documenting the new function, as we'll probably not make it public, but keep it internal. Just document the new parameter.

**Comment by RohithS98** [comment]
Sure. For the test cases, should I add a new file or just add them to tests/unit/igraph_bipartite_game.c?

**Comment by szhorvat** [comment]
Adding them to tests/unit/igraph_bipartite_game.c will be fine.

**Comment by RohithS98** [comment]
I also noticed that for `igraph_bipartite_game_gnm` directed test cases, the value of m hadn't been explicitly set before running the test, so all the tests are using `m = 0`. (Around line 93). Should I initialize m to some value or just leave it as is?

File: src/misc/bipartite.c [hunk 1]
```diff
@@ -1084,6 +1084,80 @@ igraph_error_t igraph_bipartite_game_gnp(igraph_t *graph, igraph_vector_bool_t *
...
+    IGRAPH_CHECK(igraph_vector_int_reserve(&edges, m * 2));
+
+    RNG_BEGIN();
+    for (i = 0; i < m; i++) {
+        igraph_integer_t to, from;
+
+        to = RNG_INTEGER(n1, n - 1);
+        from = RNG_INTEGER(0, n1 - 1);
+
+        if (mode == IGRAPH_IN || (mode == IGRAPH_ALL && RNG_INTEGER(0,1) == 1)) {   //Flip with probability 0.5 for IGRAPH_ALL
```
Comment (szhorvat) [inline src/misc/bipartite.c:1142]:
```suggestion
        if (mode == IGRAPH_IN || (mode == IGRAPH_ALL && RNG_INTEGER(0,1) == 1)) { /* flip with probability 0.5 for IGRAPH_ALL */
```

File: src/misc/bipartite.c [hunk 2]
```diff
@@ -1152,6 +1228,11 @@ igraph_error_t igraph_bipartite_game_gnm(igraph_t *graph, igraph_vector_bool_t *
         }
         IGRAPH_CHECK(igraph_empty(graph, n, directed));
     } else {
+
+        if (multiple){
```
Comment (szhorvat) [inline src/misc/bipartite.c:1232]:
```suggestion
        if (multiple) {
```

File: src/misc/bipartite.c [hunk 3]
```diff
@@ -1084,6 +1084,80 @@ igraph_error_t igraph_bipartite_game_gnp(igraph_t *graph, igraph_vector_bool_t *
...
+ * \return Error code.
+ *
+ * \sa \ref igraph_erdos_renyi_game_gnm() for the unipartite version,
+ * \ref igraph_bipartite_game_gnp() for the G(n1, n2, p) model.
+ *
+ * Time complexity: O(|V|+|E|), linear in the number of vertices and
+ * edges.
+ */
+
+igraph_error_t igraph_bipartite_game_gnm_multi(igraph_t *graph, igraph_vector_bool_t *types,
```
Comment (szhorvat) [inline src/misc/bipartite.c:1122]:
```suggestion
static igraph_error_t igraph_bipartite_game_gnm_multi(igraph_t *graph, igraph_vector_bool_t *types,
```

**Review by szhorvat — APPROVED** [review]

**Comment by szhorvat** [comment]
> I also noticed that for `igraph_bipartite_game_gnm` directed test cases, the value of m hadn't been explicitly set before running the test, so all the tests are using `m = 0`. (Around line 93). Should I initialize m to some value or just leave it as is?

Thanks for pointing this out. This was an oversight on my part. It would be great to fix this, but that should be a separate PR targeted to the `master` branch.

**Comment by szhorvat** [comment]
Thanks for your work on this @RohithS98! It looks great. I'll let @ntamas merge if he's also happy.

In the meantime, can you please update `CHANGELOG.md`, and mention that new parameter?

**Review by ntamas — APPROVED** [review]

**Comment by ntamas** [comment]
Thanks a lot!

**Pull request merged**